bluepill/src/BPPacker.m

```diff
@@ -86,5 +96,74 @@ @implementation BPPacker
     return bundles;
 }
 
++ (NSMutableArray<BPXCTestFile *> *)packTestsByTime:(NSArray<BPXCTestFile *> *)xcTestFiles
+                                      configuration:(BPConfiguration *)config
+                                           andError:(NSError **)errPtr {
+    [BPUtils printInfo:INFO withString:@"Packing based on individual test execution times in file path: %@", config.testTimeEstimatesJsonFile];
+    if (xcTestFiles.count == 0) {
+        BP_SET_ERROR(errPtr, @"Found no XCTest files.\n"
+                     "Perhaps you forgot to 'build-for-testing'? (Cmd + Shift + U) in Xcode.");
+        return NULL;
+    }
+
+    // load the config file
+    NSError *error;
+    NSDictionary *testTimes = [self loadConfigFile:config.testTimeEstimatesJsonFile withError:&error];
+    if (error || !testTimes) {
+        BP_SET_ERROR(errPtr, @"Could not load configuration from '%@'\n%@", optarg, [error localizedDescription]);
+        return NULL;
+    }
+
+    NSMutableDictionary *estimatedTimesByTestFilePath = [[NSMutableDictionary alloc] init];
+    NSArray *testCasesToRun = config.testCasesToRun;
+
+    double totalTime = 0.0;
+    for (BPXCTestFile *xctFile in xcTestFiles) {
+        NSMutableSet *bundleTestsToRun = [[NSMutableSet alloc] initWithArray:[xctFile allTestCases]];
+        if (testCasesToRun) {
+            [bundleTestsToRun intersectSet:[[NSSet alloc] initWithArray:testCasesToRun]];
+        }
+        if (config.testCasesToSkip && [config.testCasesToSkip count] > 0) {
+            [bundleTestsToRun minusSet:[[NSSet alloc] initWithArray:config.testCasesToSkip]];
+        }
+        if (bundleTestsToRun.count > 0) {
+            double __block testBundleExecutionTime = 0.0;
+            [bundleTestsToRun enumerateObjectsUsingBlock:^(id _Nonnull test, BOOL * _Nonnull stop) {
+                // TODO: Assign a sensible default if the estimate is not given
```

Comment:
Do you want to keep this TODO?
If it's 0 multiple new test case maybe added to the same bundle without impacting the estimation.
Maybe use the timeout or a medium number from the estimation json?

Reply:
@chenxiao0228, Right now some non-tests are being considered as tests -- which I think is a bug. Using the median was the plan, once this BP parsing bug is fixed.